ENH: Added submodule numpy.random_intel

oleksandr-pavlyk opened the pull request:
`numpy.random_intel` is MKL-based alternative for `numpy.random`, first
featured in the [Intel Distribution for Python](https://software.intel.com/en-us/intel-distribution-for-python).

It implements all distributions of `numpy.random` as well as
`numpy.random_intel.multinormal_cholesky`, which allows for fast
sampling from multinormal distribution when the Cholesky factor
is known ahead of time.

Module `numpy.random_intel` implements sampling from all distributions
in a vectorized fashion, just like MKL does, which allows MKL to use
vectorized instructions for efficiency. 

That means that instead of mtrand's `rk_normal` which produces a single 
variate sampled from normal distribution,  mklrand offers `vrk_normal_vec`, 
which produces a vector of variates of requested size.

EDIT: link to mailing list design discussion on this feature: https://mail.scipy.org/pipermail/numpy-discussion/2016-July/075822.html

oleksandr-pavlyk committed:
ENH: Added submodule numpy.random_intel

numpy.random_intel is MKL-based alternative for numpy.random, first
featured in the Intel Distribution for Python.

It implements all distributions of numpy.random as well as
numpy.random_intel.multinormal_cholesky, which allows for fast
sampling from multinormal distribution when the Cholseky factor
is known ahead of time.

Module numpy.random_intel implements sampling from all distributions
in a vectorized fashion, just like MKL does, which allows MKL to use
vectorized instructions for efficiency. That means that instead of
mtrand's rk_normal which produces a single variate sampled from normal
distribution,  mklrand offers vrk_normal_vec, which produces a vector of
variates of requested size.

oleksandr-pavlyk commented:
@njsmith @rkern

charris commented on numpy/random_intel/mklrand/Python.pxi:
```diff
@@ -0,0 +1,58 @@
+    void Py_XINCREF(object obj)
+
+    # CObject API
+# If this is uncommented it needs to be fixed to use PyCapsule
```
Take a look at `numpy/core/include/numpy/npy_3kcompat.h`. Given the current organization, we might want to move that or this work. Not sure what the best organozation will be.

charris replied:
Also note that `PyCapsule` is available in Python 2.7 and we no longer support earlier versions. In fact, PyCapsule has been used in `numpy/random/mtrand/mtrand.pyx` since 1.11.x was branched.

oleksandr-pavlyk replied:
I copied `mtrand/Python.pxi` over to `mklrand/Python.pxi'. Thanks!

charris commented:
What I notice at first glance are long lines (we like < 80 char/line) and not much documentation.  There is `doc/C_STYLE_GUIDE.rst.txt` for numpy C code, and PEP8 for Python. I'm not sure of the best way to document without too much repetition of current random function documentation. I'm also  not sure of the best way will be to handle two distinct modules covering the same material, or whether you should put this work up on PyPI as a separately loadable module -- which was pretty much my first thought.

It would be good to bring this up on the numpy mailing list for discussion: numpy-discussion@scipy.org. @rkern, @njsmith Thoughts?

rgommers commented:
I've edited the PR description to link to the mailing list discussion in July on this. 

As far as I can judge from a browse of this PR, this implementation is still standalone and doesn't reflect the design @rkern said he preferred right?

oleksandr-pavlyk committed:
BUG: Tests in numpy/random_intel/tests are to be skipped if
numpy has not been configured with MKL.

Added special case of warning 'ignore' detection to skip
random_intel/__init__.py, parallel to skipping random/__init__.py

oleksandr-pavlyk committed:
Copied mtrand/Python.pxi to mklrand/Python.pxi to
address PR feedback

charris commented:
On Mon, Oct 24, 2016 at 4:03 PM, Oleksandr Pavlyk <notifications@github.com>
wrote:

> @njsmith <https://github.com/njsmith> @rkern <https://github.com/rkern>
>

Hmm, I don't see any replies so far. Other projects that might be
interested are scikit-learn and stats-models.

Chuck

oleksandr-pavlyk committed:
formatting adjustments to mkl_distributions.cpp to comply with C-style guidelines adopted by numpy

njsmith commented:
@oleksandr-pavlyk: So you know how this works... usually we use pull request comments to talk about details of code, and for higher-level discussions that affect public API (like: "should numpy have a `random_intel` module at all?") we switch to the mailing list. That discussion is happening there now, and you probably want to read and participate :-). The mailing list thread starts here: https://mail.scipy.org/pipermail/numpy-discussion/2016-October/076144.html

rkern commented:
> Hmm, I don't see any replies so far. Other projects that might be
> interested are scikit-learn and stats-models.

I've made my thoughts clear on the mailing list, in this incarnation and
the previous thread. This should be in a separate package.

charris commented:
@oleksandr-pavlyk Sounds like a separate package it is, so closing this. Thanks for your work on this.
